### PR TITLE
feat: add new token selector and  show balance

### LINKED ISSUE #2926: [Demand] Display token balance in the select token dialog
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

image

## Part 2: Problem to Solve

image

## Notes

[Uniswap Interface](https://github.com/Uniswap/uniswap-interface) is a good reference.

### PR DESCRIPTION
**Pull request by NikoManiac**
closes #MASK-51
closes #2926

### CODE REVIEW

**Comment by septs** [comment]
@NikoManiac `packages/dashboard/src/components/TokenIcon/TokenIconSpeialIconList.json` the file name typo error
`Speial` -> `Special`

File: packages/dashboard/tsconfig.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
     "rootDir": "./",
     "outDir": "./dist/"
   },
-  "include": ["./src/", "./stories/", "./src/locales/*.json"],
+  "include": ["./src/", "./stories/", "./src/**/*.json"],
```
Comment (septs) [inline packages/dashboard/tsconfig.json:7]: why

**Comment by NikoManiac** [comment]
> @NikoManiac `packages/dashboard/src/components/TokenIcon/TokenIconSpeialIconList.json` the file name typo error
> `Speial` -> `Special`

I will rename this file,  thanks.

File: packages/dashboard/tsconfig.json (hunk 1 above)
Comment (NikoManiac) [reply packages/dashboard/tsconfig.json:7]: I need to import TokenIconSpecialIconList.json

File: packages/dashboard/src/components/TokenIcon/index.tsx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    const fullIconAssetURIs = baseURIs.map((x) => `${x}/assets/${checkSummedAddress}/logo.png`)
+    return logoURI ? [logoURI] : fullIconAssetURIs
+}
+//#endregion
+
+const useStyles = makeStyles((theme: Theme) => ({
+    icon: {
+        width: 36,
+        height: 36,
+        backgroundColor: theme.palette.common.white,
```
Comment (yanzhihong23) [inline packages/dashboard/src/components/TokenIcon/index.tsx:38]: white background in both light and dark mode?

Comment (NikoManiac) [reply packages/dashboard/src/components/TokenIcon/index.tsx:38]: I tested on local. It's ok for token icon.

File: packages/theme/src/Components/SearchableList/SearchableList.tsx [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        paddingBottom: theme.spacing(4),
+    },
+    textField: {
+        '&>div': {
+            borderRadius: theme.spacing(1),
+            background: MaskColorVar.normalBackground,
+        },
+    },
+    searchInput: {
+        padding: '-6px',
```
Comment (UncleBill) [inline packages/theme/src/Components/SearchableList/SearchableList.tsx:134]: `padding` doesn't support negative values, do you mean `6px`?

File: packages/dashboard/src/components/TokenIcon/index.tsx [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+//#region fix icon image
+function resolveTokenIconURLs(address: string, baseURIs: string[], chainId: ChainId, logoURI?: string) {
+    const checkSummedAddress = formatEthereumAddress(address)
+
+    if (isSameAddress(constantOfChain(TOKEN_CONSTANTS, chainId).NATIVE_TOKEN_ADDRESS, checkSummedAddress)) {
+        return baseURIs.map((x) => `${x}/info/logo.png`)
+    }
+
+    const specialIcon = SPECIAL_ICON_LIST.find((item) => item.address === address)
```
Comment (guanbinrui) [inline packages/dashboard/src/components/TokenIcon/index.tsx:25]: `currySameToken`

File: packages/theme/src/Components/SearchableList/SearchableList.tsx (hunk 3 above)
Comment (NikoManiac) [reply packages/theme/src/Components/SearchableList/SearchableList.tsx:134]: This is not valid, I will remove it.

File: packages/dashboard/src/components/TokenIcon/index.tsx (hunk 4 above)
Comment (NikoManiac) [reply packages/dashboard/src/components/TokenIcon/index.tsx]: done.

**Review by guanbinrui — APPROVED** [review]

File: packages/dashboard/src/components/TokenIcon/TokenIconSpecialIconList.json [hunk 5]
```diff
@@ -0,0 +1,167 @@
+[
+    {
```
Comment (Jack-Works) [inline packages/dashboard/src/components/TokenIcon/TokenIconSpecialIconList.json:2]: Should we need #3452 for this change?

Comment (guanbinrui) [reply packages/dashboard/src/components/TokenIcon/TokenIconSpecialIconList.json:2]: It's a specific purpose JSON file.

File: packages/dashboard/src/components/TokenIcon/index.tsx [hunk 6]
```diff
@@ -0,0 +1,70 @@
+import { makeStyles, Avatar, Theme, AvatarProps } from '@material-ui/core'
```
Comment (nuanyang233) [inline packages/dashboard/src/components/TokenIcon/index.tsx:1]: Here there is already a TokenIcon component in `src/pages/wallet/TokenIcon`.

File: packages/dashboard/src/pages/Wallets/hooks/useERC20TokensDetailed.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
+import {
```
Comment (nuanyang233) [inline packages/dashboard/src/pages/Wallets/hooks/useERC20TokensDetailed.ts:1]: 🤔Why doesn't it move to web3-shared? And I think the existing Hook should be able to meet your needs and there should be no need to create a new hook

Comment (NikoManiac) [reply packages/dashboard/src/pages/Wallets/hooks/useERC20TokensDetailed.ts:1]: I think this hook is work for dashboard UI. It wrapping the useERC20TokensDetailedFromTokenLists, an improved TokenList component code readability.

File: packages/dashboard/src/components/TokenList/useImageFailover.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import { useAsync } from 'react-use'
+import { findAvailableImageURL } from '@dimensiondev/kit'
+
+export function useImageFailover(urls: readonly string[], suffix: string) {
```
Comment (Jack-Works) [inline packages/dashboard/src/components/TokenList/useImageFailover.ts:4]: Is this function copied from anywhere in our codebase? It looks familiar.

File: packages/theme/src/Components/SearchableList/SearchableList.tsx [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+import { InputAdornment, makeStyles, TextField, Typography } from '@material-ui/core'
+import { Search } from '@material-ui/icons'
+import { MaskColorVar } from '../../constants'
+import { MaskSearchableItemInList } from './MaskSearchableItemInList'
+
+export interface MaskSearchableListProps<T> {
+    /** The list data should be render */
+    data: T[]
+    /** The identity of list data item for remove duplicates item */
+    key?: keyof T
```
Comment (Jack-Works) [inline packages/theme/src/Components/SearchableList/SearchableList.tsx:14]: The `key` props will be removed by React.

**Comment by septs** [comment]
plz resolve conflict.

File: packages/shared/src/UI/components/TokenIcon/index.tsx [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    address: string
+    AvatarProps?: Partial<AvatarProps>
+}
+
+export function TokenIcon(props: TokenIconProps) {
+    const { address, logoURI, name, chainId, AvatarProps, classes } = props
+
+    const chainDetailed = useChainDetailed()
+    const tokenBlockie = useBlockie(address)
+    const tokenAssetBaseURI = getTokenAssetBaseURLConstants().TOKEN_ASSET_BASE_URI
```
Comment (septs) [inline packages/shared/src/UI/components/TokenIcon/index.tsx:56]:
```suggestion
    const { TOKEN_ASSET_BASE_URI } = getTokenAssetBaseURLConstants()
```

**Comment by NikoManiac** [comment]
> @NikoManiac hi please split the refactor (moving things) to another PR so it will be merged first.

The ErrorBoundary / QRCode in theme package which depends on shared package. The shared package is depends on  theme package such as SearchableList. This is a circular dependency.

So I decided to deal with the dependency first.

**Comment by NikoManiac** [comment]
@Jack-Works hi jack, 
The ErrorBoundary / QRCode in theme package which depends on shared package. The shared package is depends on theme package such as SearchableList. This is a circular dependency.

So I decided to deal with the dependency first.

File: packages/dashboard/src/pages/Wallets/hooks/useERC20TokensDetailed.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+import { uniqBy } from 'lodash-es'
+import { useMemo } from 'react'
+
+export function useERC20TokensDetailed() {
+    const ERC20_TOKEN_LISTS = getEthereumConstants().ERC20_TOKEN_LISTS
+    const { value: nativeTokenDetailed } = useNativeTokenDetailed()
+    const { state, tokensDetailed: erc20TokensDetailed } = useERC20TokensDetailedFromTokenLists(ERC20_TOKEN_LISTS)
+
+    //#region mask token
+    const MASK_ADDRESS = getTokenConstants().MASK_ADDRESS
```
Comment (septs) [inline packages/dashboard/src/pages/Wallets/hooks/useERC20TokensDetailed.ts:20]:
```suggestion
    const { MASK_ADDRESS } = getTokenConstants()
```

Comment (NikoManiac) [reply packages/dashboard/src/pages/Wallets/hooks/useERC20TokensDetailed.ts]: done

**Review by guanbinrui — APPROVED** [review]

File: packages/maskbook/src/settings/settings.ts [hunk 12]
```diff
@@ -52,12 +53,27 @@ export const enableGroupSharingSettings = createGlobalSettings<boolean>('experim
...
+ * Expected Usage：export const currentImagePayloadStatus = createNetworkSettings('currentImagePayloadStatus')
+ *
+ * Work around the issue:
+ *      https://github.com/microsoft/TypeScript/issues/42873
+ *      https://github.com/microsoft/TypeScript/issues/30858
+ *
+ * References:
+ *      PluginGitcoinMessages: packages/maskbook/src/plugins/Gitcoin/messages.ts
+ *      PluginTraderMessages: packages/maskbook/src/plugins/Trader/messages.ts
+ *      PluginTransakMessages: packages/maskbook/src/plugins/Transak/messages.ts
```
Comment (Jack-Works) [inline packages/maskbook/src/settings/settings.ts:68]: IMO no need to list usages

File: packages/shared/src/UI/components/TokenIcon/index.tsx [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+    if (isSameAddress(getTokenConstants().NATIVE_TOKEN_ADDRESS, checkSummedAddress)) {
+        return baseURIs.map((x) => `${x}/info/logo.png`)
+    }
+
+    const specialIcon = SPECIAL_ICON_LIST.find(currySameAddress(address))
+
+    if (specialIcon) return [specialIcon.logo_url]
+
+    const fullIconAssetURIs = baseURIs.map((x) => `${x}/assets/${checkSummedAddress}/logo.png`)
```
Comment (Jack-Works) [inline packages/shared/src/UI/components/TokenIcon/index.tsx:29]: Wait, are they local assets or remote assets? Please document it in the code.

File: packages/shared/src/UI/components/TokenList/index.tsx [hunk 14]
```diff
@@ -0,0 +1,37 @@
+import React, { memo, useEffect, useState } from 'react'
+import { TokenListItem } from './TokenListItem'
+import { some } from 'lodash-es'
+import type { Asset, FungibleTokenDetailed } from '@masknet/web3-shared'
+import { EthereumTokenType, isSameAddress, useERC20Tokens } from '@masknet/web3-shared'
+import { SearchableList } from '@masknet/theme'
+
+interface IProps {
```
Comment (Jack-Works) [inline packages/shared/src/UI/components/TokenList/index.tsx:8]: Rename to `TokenListProps` and export it.

**Review by septs — DISMISSED** [review]

File: packages/maskbook/src/settings/settings.ts (hunk 12 above)
Comment (NikoManiac) [reply packages/maskbook/src/settings/settings.ts:68]: I wanted remind others that these files were changed for this same issue. If need to be removed later, those files should be together.

File: packages/shared/src/UI/components/TokenIcon/index.tsx (hunk 13 above)
Comment (NikoManiac) [reply packages/shared/src/UI/components/TokenIcon/index.tsx:29]: This is remote assets, I will leave comment in the code.

**Review by Jack-Works — APPROVED** [review]

**Review by yanzhihong23 — APPROVED** [review]

**Pull request merged**